storage-api/src/java/org/apache/hadoop/hive/ql/exec/vector/BytesColumnVector.java

```diff
@@ -121,30 +125,30 @@ public void setRef(int elementNum, byte[] sourceBuf, int start, int length) {
    * Provide the estimated number of bytes needed to hold
    * a full column vector worth of byte string data.
    *
-   * @param estimatedValueSize  Estimated size of buffer space needed
+   * @param estimatedValueSize  Estimated size of buffer space needed per row
    */
   public void initBuffer(int estimatedValueSize) {
-    nextFree = 0;
     smallBufferNextFree = 0;
 
     // if buffer is already allocated, keep using it, don't re-allocate
-    if (buffer != null) {
+    if (smallBuffer != null) {
       // Free up any previously allocated buffers that are referenced by vector
       if (bufferAllocationCount > 0) {
         for (int idx = 0; idx < vector.length; ++idx) {
           vector[idx] = null;
           length[idx] = 0;
         }
-        buffer = smallBuffer; // In case last row was a large bytes value
       }
     } else {
       // allocate a little extra space to limit need to re-allocate
-      int bufferSize = this.vector.length * (int)(estimatedValueSize * EXTRA_SPACE_FACTOR);
+      long bufferSize = (long) (this.vector.length * estimatedValueSize * EXTRA_SPACE_FACTOR);
       if (bufferSize < DEFAULT_BUFFER_SIZE) {
         bufferSize = DEFAULT_BUFFER_SIZE;
       }
-      buffer = new byte[bufferSize];
-      smallBuffer = buffer;
+      if (bufferSize > MAX_SIZE_FOR_SMALL_BUFFER) {
```

Comment:
is this condition valid? Would we have enough space to store the new element when > MAX_SIZE_FOR_SMALL_BUFFER?

Reply:
In initBuffer, we are using an estimate of the value size to get the initial size. If a given value is larger than MAX_SIZE_FOR_SMALL_ITEM it will be put into a temporary buffer. This check makes sure that we don't use a large estimate to allocate a huge backing buffer.